tour_model_eval/label_processing.py

```diff
@@ -0,0 +1,98 @@
+import logging
+from scipy.cluster.hierarchy import linkage, dendrogram,fcluster
+
+
+
+# to map the user labels
+# - user_input_df: pass in original user input dataframe, return changed user input dataframe
+# - sp2en: change Spanish to English
+# - cvt_pur_mo: convert purposes and replaced mode
+def map_labels(user_input_df,sp2en,cvt_pur_mo):
+    # Spanish words to English
+    span_eng_dict = {'revisado_bike': 'test ride with bike', 'placas_de carro': 'car plates', 'aseguranza': 'insurance',
+                     'iglesia': 'church', 'curso': 'course',
+                     'mi_hija recién aliviada': 'my daughter just had a new baby',
+                     'servicio_comunitario': 'community service', 'pago_de aseguranza': 'insurance payment',
+                     'grupo_comunitario': 'community group', 'caminata_comunitaria': 'community walk'}
+
+    # Convert purpose
+    map_pur_dict = {'course': 'school', 'work_- lunch break': 'lunch_break', 'on_the way home': 'home',
+                    'insurance_payment': 'insurance'}
+
+    if sp2en:
+        # change language
+        user_input_df = user_input_df.replace(span_eng_dict)
+    elif cvt_pur_mo:
+        # change language first
+        user_input_df = user_input_df.replace(span_eng_dict)
+        # convert purpose
+        user_input_df = user_input_df.replace(map_pur_dict)
+        # convert mode
+        for a in range(len(user_input_df)):
+            if user_input_df.iloc[a]["replaced_mode"] == "same_mode":
+                # to see which row will be converted
+                logging.debug("The following rows will be changed: %s", user_input_df.iloc[a])
+                user_input_df.iloc[a]["replaced_mode"] = user_input_df.iloc[a]['mode_confirm']
+    return user_input_df
+
+
+# use hierarchical clustering to get labels of the second round
+def get_second_labels(x,method,low,dist_pct):
+    z = linkage(x, method=method, metric='euclidean')
+    last_d = z[-1][2]
+    clusters = []
+    if last_d < low:
+        for i in range(len(x)):
+            clusters.append(0)
+    else:
+        max_d = last_d * dist_pct
+        clusters = fcluster(z, max_d, criterion='distance')
+    return clusters
+
+
+# this function includes hierarchical clustering and changing labels to get appropriate labels for
+# the second round of clustering
+# appropriate labels are label from the first round concatenate label from the second round
+# (e.g. label from first round is 1, label from second round is 2, the new label will be 12)
+def get_new_labels(x,low,dist_pct,second_round_idx_labels,new_labels,method=None):
+    idx_label = second_round_idx_labels.copy()
```

Comment:
what is `second_round_idx_labels` and why do you need to `copy()` it?

Reply:
It is a track to store the indices and labels for the second round.